plugins/inputs/mavlink/mavlink_utils.go

```diff
@@ -0,0 +1,130 @@
+package mavlink
+
+import (
+	"fmt"
+	"net"
+	"net/url"
+	"reflect"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/chrisdalke/gomavlib/v3"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/metric"
+)
+
+// Convert a string from CamelCase to snake_case
+// There is no single convention for Mavlink message names - Sometimes
+// they are referenced as CAPITAL_SNAKE_CASE. Gomavlink converts them
+// to CamelCase. This plugin takes an opinionated stance and makes the
+// message names and field names all lowercase snake_case.
+func ConvertToSnakeCase(input string) string {
+	re := regexp.MustCompile(`([a-z0-9])([A-Z])`)
+	snake := re.ReplaceAllString(input, `${1}_${2}`)
+	snake = strings.ToLower(snake)
+	return snake
+}
+
+// Convert a Mavlink event into a struct containing Metric data.
+func MavlinkEventFrameToMetric(frm *gomavlib.EventFrame) telegraf.Metric {
+	m := frm.Message()
+	t := reflect.TypeOf(m)
+	v := reflect.ValueOf(m)
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+		v = v.Elem()
+	}
+
+	messageName := ConvertToSnakeCase(t.Name())
+	messageName = strings.TrimPrefix(messageName, "message_")
+
+	out := metric.New(
+		messageName,
+		make(map[string]string),
+		make(map[string]interface{}),
+		time.Unix(0, 0),
+	)
+	out.AddTag("sys_id", strconv.FormatUint(uint64(frm.SystemID()), 10))
+
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+		value := v.Field(i)
+		out.AddField(ConvertToSnakeCase(field.Name), value.Interface())
+	}
+
+	return out
+}
```

Comment:
Please do not export this function. Furthermore, preallocating the tags and fields saves some memory trashing
```suggestion
func convertEventFrameToMetric(frm *gomavlib.EventFrame) telegraf.Metric {
	m := frm.Message()
	t := reflect.TypeOf(m)
	v := reflect.ValueOf(m)
	if t.Kind() == reflect.Ptr {
		t = t.Elem()
		v = v.Elem()
	}

	name = internal.SnakeCase(strings.TrimPrefix(t.Name(), "Message"))
	tags := map[string]string{
		"sys_id", strconv.FormatUint(uint64(frm.SystemID()), 10),
	}
	fields := make(map[string]interface{}, t.NumField())

	for i := 0; i < t.NumField(); i++ {
		field := t.Field(i)
		value := v.Field(i)
		fields[internal.SnakeCase(field.Name)] = value.Interface())
	}
	
	return 	out := metric.New(name, tags, fields, time.Now())
}
```